pkg/apis/autoscaling/v2beta1/conversion.go

```diff
@@ -0,0 +1,399 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v2beta1
+
+import (
+	"encoding/json"
+
+	autoscalingv2beta1 "k8s.io/api/autoscaling/v2beta1"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kubernetes/pkg/apis/autoscaling"
+	core "k8s.io/kubernetes/pkg/apis/core"
+)
+
+func addConversionFuncs(scheme *runtime.Scheme) error {
+	// Add non-generated conversion functions
+	err := scheme.AddConversionFuncs(
+		Convert_autoscaling_ExternalMetricSource_To_v2beta1_ExternalMetricSource,
+		Convert_v2beta1_ExternalMetricSource_To_autoscaling_ExternalMetricSource,
+		Convert_autoscaling_ObjectMetricSource_To_v2beta1_ObjectMetricSource,
+		Convert_v2beta1_ObjectMetricSource_To_autoscaling_ObjectMetricSource,
+		Convert_autoscaling_PodsMetricSource_To_v2beta1_PodsMetricSource,
+		Convert_v2beta1_PodsMetricSource_To_autoscaling_PodsMetricSource,
+		Convert_autoscaling_ExternalMetricStatus_To_v2beta1_ExternalMetricStatus,
+		Convert_v2beta1_ExternalMetricStatus_To_autoscaling_ExternalMetricStatus,
+		Convert_autoscaling_ObjectMetricStatus_To_v2beta1_ObjectMetricStatus,
+		Convert_v2beta1_ObjectMetricStatus_To_autoscaling_ObjectMetricStatus,
+		Convert_autoscaling_PodsMetricStatus_To_v2beta1_PodsMetricStatus,
+		Convert_v2beta1_PodsMetricStatus_To_autoscaling_PodsMetricStatus,
+		Convert_autoscaling_ResourceMetricSource_To_v2beta1_ResourceMetricSource,
+		Convert_v2beta1_ResourceMetricSource_To_autoscaling_ResourceMetricSource,
+		Convert_autoscaling_MetricTarget_To_v2beta1_CrossVersionObjectReference,
+		Convert_v2beta1_CrossVersionObjectReference_To_autoscaling_MetricTarget,
+		Convert_autoscaling_ResourceMetricStatus_To_v2beta1_ResourceMetricStatus,
+		Convert_v2beta1_ResourceMetricStatus_To_autoscaling_ResourceMetricStatus,
+		Convert_autoscaling_HorizontalPodAutoscaler_To_v2beta1_HorizontalPodAutoscaler,
+		Convert_v2beta1_HorizontalPodAutoscaler_To_autoscaling_HorizontalPodAutoscaler,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func Convert_autoscaling_MetricTarget_To_v2beta1_CrossVersionObjectReference(in *autoscaling.MetricTarget, out *autoscalingv2beta1.CrossVersionObjectReference, s conversion.Scope) error {
```

Comment:
why are these nil?  If they're not needed, don't implement them

Reply:
That's what I thought, but without them defined as nil they get generated for some reason:

```
# k8s.io/kubernetes/pkg/apis/autoscaling/v2beta1
pkg/apis/autoscaling/v2beta1/zz_generated.conversion.go:507:12: undefined: Convert_v2beta1_CrossVersionObjectReference_To_autoscaling_MetricTarget
pkg/apis/autoscaling/v2beta1/zz_generated.conversion.go:519:12: undefined: Convert_autoscaling_MetricTarget_To_v2beta1_CrossVersionObjectReference
!!! [0530 21:24:29] Call tree:
!!! [0530 21:24:29]  1: /home/mdame/go/src/k8s.io/kubernetes/hack/lib/golang.sh:595 kube::golang::build_binaries_for_platform(...)
!!! [0530 21:24:29]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0530 21:24:29] Call tree:
!!! [0530 21:24:29]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0530 21:24:29] Call tree:
!!! [0530 21:24:29]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:92: all] Error 1
```

I think it has to do with the fact that `Target` in `v2beta1` is a `CrossVersionObjectReference`, but in `v2beta2` we renamed that to `DescribedObject` and created a *new* field called `Target` that is a `MetricTarget`. So it's seeing Target=Target and trying to auto-convert CrossVersionObjectReference->MetricTarget and telling me I need to manually convert it (which I can't)

If I'm wrong, or there's a way around that, I can make the change (or I can add a comment explaining these functions, which I probably should have done first)